### PR TITLE
Refactored FTP

### PR DESCRIPTION
**Pull request by johnathan-rodriguez**

### CODE REVIEW

File: shipyard_blueprints/ftp/pyproject.toml [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
 license = "Apache 2.0"
 readme = "README.md"
-packages = [{include = "shipyard_ftp"}]
+packages = [{ include = "shipyard_ftp" }]
 
 [tool.poetry.dependencies]
-python = ">=3.7,<3.10"
-shipyard-templates = "0.5.0a0"
-shipyard-utils = "0.1.4"
+python = ">=3.9,<3.10"
```
Comment (wrp801) [inline shipyard_blueprints/ftp/pyproject.toml:11]: the python version should be: 
`python = "^3.9"`

File: shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py [hunk 2]
```diff
@@ -29,115 +32,63 @@ def get_args():
...
-
-        if number_of_matches == 0:
-            print(f'No matches were found for regex "{file_name}".')
-            sys.exit(ec.EXIT_CODE_NO_MATCHES_FOUND)
-
-        for index, file_name in enumerate(matching_file_names):
-            print(
-                f"Deleting file {index+1} of {len(matching_file_names)} out of {number_of_matches}"
-            )
+            number_of_matches = len(matching_file_names)
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py:62]: I kind of like the `if (n_matches := len(matching_file_names)) == 0` paradigm

File: shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py [hunk 3]
```diff
@@ -29,115 +32,63 @@ def get_args():
...
+
+            if number_of_matches == 0:
+                logger.info(f'No matches were found for regex "{file_name}".')
+                sys.exit(EXIT_CODE_NO_MATCHES_FOUND)
+
+            for index, file_name in enumerate(matching_file_names):
+                logger.info(
+                    f"Deleting file {index + 1} of {len(matching_file_names)} out of {number_of_matches}"
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py:70]: This should be: 
` f"Deleting file {index + 1} of {number_of_matches}"`

File: shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py [hunk 4]
```diff
@@ -29,115 +32,63 @@ def get_args():
...
-        try:
-            delete_ftp_file(client=client, file_path=file_path)
-        except Exception as e:
-            print(f"The server says: {e}")
-            print(
-                f"Most likely, the file name/folder name you specified has typos or the full folder name was not provided. Check these and try again."
-            )
-            sys.exit(ec.EXIT_CODE_NO_MATCHES_FOUND)
+                logger.error(
+                    "Most likely, the file name/folder name you specified has typos or the full folder name was not "
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py:82]: I would rephrase this error message as:

"Check the file/folder name for misspellings and ensure that entire folder name was provided"

Maybe set the level to `warning` as well?

File: shipyard_blueprints/ftp/shipyard_ftp/cli/download.py [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
-import ftplib
+from shipyard_bp_utils import files as shipyard
+from shipyard_templates import ExitCodeException, CloudStorage
+from shipyard_templates.shipyard_logger import ShipyardLogger
+
+from shipyard_ftp.ftp import FtpClient
+
+logger = ShipyardLogger().get_logger()
 
 EXIT_CODE_INCORRECT_CREDENTIALS = 3
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/download.py:13]: I think these should be relocated to the `exceptions.py` file

File: shipyard_blueprints/ftp/shipyard_ftp/cli/download.py [hunk 6]
```diff
@@ -171,126 +58,88 @@ def find_matching_files(file_names, file_name_re):
...
-            files, folders = find_files_in_directory(
-                client=client, folder_filter=folder_filter, files=files, folders=folders
+            matching_file_names = shipyard.find_all_file_matches(
+                files, re.compile(source_file_name)
             )
 
-        matching_file_names = find_matching_files(files, re.compile(source_file_name))
-
-        number_of_matches = len(matching_file_names)
+            number_of_matches = len(matching_file_names)
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/download.py:112]: same comment as above using the `:=` operator inline with the `if` statement. Not a requirement, I just think it adds a bit of syntactic sugar

File: shipyard_blueprints/ftp/shipyard_ftp/cli/download.py [hunk 7]
```diff
@@ -171,126 +58,88 @@ def find_matching_files(file_names, file_name_re):
...
-                destination_file_name=destination_name,
-            )
-        except Exception as e:
-            print(f"The server says: {e}")
-            print(
-                "Most likely, the file name/folder name you specified has typos or the full folder name was not provided. Check these and try again."
-            )
-            sys.exit(EXIT_CODE_NO_MATCHES_FOUND)
+                logger.info(
+                    f"Downloading file {index + 1} of {len(matching_file_names)}"
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/download.py:131]: Put the destination path in the log as well

File: shipyard_blueprints/ftp/shipyard_ftp/cli/download.py [hunk 8]
```diff
@@ -171,126 +58,88 @@ def find_matching_files(file_names, file_name_re):
...
+                f"{len(matching_file_names)} files found. Preparing to download..."
             )
 
-            print(f"Downloading file {index + 1} of {len(matching_file_names)}")
-            try:
-                download_ftp_file(
-                    client=client,
-                    file_name=file_name,
-                    destination_file_name=destination_name,
+            for index, file_name in enumerate(matching_file_names):
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/download.py:122]: add `start = 1` in enumerate so you don't have to do `index + 1` below

File: shipyard_blueprints/ftp/shipyard_ftp/cli/download.py [hunk 9]
```diff
@@ -171,126 +58,88 @@ def find_matching_files(file_names, file_name_re):
...
-            )
-            sys.exit(EXIT_CODE_NO_MATCHES_FOUND)
+                logger.info(
+                    f"Downloading file {index + 1} of {len(matching_file_names)}"
+                )
 
-    print("Download process complete.")
+                client.download(file_name, destination_name)
+
+        logger.info("Download process complete.")
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/download.py:136]: I think it is useful to see in the logs where the file is being downloaded to. What about something like:
`logger.info('Downloaded file to {destination_path}')`. Again, not a requirement, but I think that provides more useful information than "Download process complete"

File: shipyard_blueprints/ftp/shipyard_ftp/cli/move.py [hunk 10]
```diff
@@ -132,67 +50,58 @@ def main():
...
+    elif source_file_name_match_type == "regex_match":
+        file_names = ftp_client.get_all_nested_items(source_folder_name)
+        matching_file_names = shipyard.find_all_file_matches(
+            file_names, re.compile(source_file_name)
+        )
 
-        if number_of_matches == 0:
-            print(f'No matches were found for regex "{source_file_name}".')
-            sys.exit(ec.EXIT_CODE_NO_MATCHES_FOUND)
+        if len(matching_file_names) == 0:
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/move.py:84]: Same comment with the `:=` to consider

File: shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py [hunk 11]
```diff
@@ -41,229 +42,87 @@ def get_args():
...
-    if destination_file_name:
-        if file_number:
-            destination_file_name = enumerate_destination_file_name(
-                destination_file_name, file_number
+        source_file_name_match_type = args.source_file_name_match_type
+        errors = []
+        ftp_client = FtpClient(host=host, port=port, user=username, pwd=password)
+        if source_file_name_match_type == "exact_match":
+            source_full_path = shipyard.combine_folder_and_file_name(
+                folder_name=f"{os.getcwd()}/{source_folder_name}",
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py:63]: Is it necessary to join the current working directory to the source folder name instead of just provide the source folder name as the argument?

File: shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py [hunk 12]
```diff
@@ -41,229 +42,87 @@ def get_args():
...
-        if len(destination_full_path.split("/")) > 1:
-            path, file_name = destination_full_path.rsplit("/", 1)
-            create_new_folders(client=client, destination_path=path)
-
-        upload_ftp_file(
-            client=client,
-            source_full_path=source_full_path,
-            destination_full_path=destination_full_path,
-        )
+            for index, key_name in enumerate(matching_file_names):
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py:100]: start the enumeration at 1

File: shipyard_blueprints/ftp/shipyard_ftp/ftp.py [hunk 13]
```diff
@@ -1,32 +1,238 @@
...
+            raise
+        except Exception as e:
+            raise exceptions.MoveError(
+                f"Failed to move {source_path}. Ensure that the source/destination file name and folder name are "
+                f"correct. Error message from server: {e} ",
+            ) from e
+
+        logger.info(f"{source_path} successfully moved to " f"{dest_path}")
+
+    def remove(self, file_path):
```
Comment (wrp801) [inline shipyard_blueprints/ftp/shipyard_ftp/ftp.py]: Add docstrings to `remove` and `upload`

**Review by wrp801 — CHANGES_REQUESTED** [review]
small quality of life changes

File: shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py (hunk 4 above)
Comment (johnathan-rodriguez) [reply shipyard_blueprints/ftp/shipyard_ftp/cli/delete.py:82]: Made the rewording change but left as error since something did fail and it raises an error

File: shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py (hunk 11 above)
Comment (johnathan-rodriguez) [reply shipyard_blueprints/ftp/shipyard_ftp/cli/upload.py:63]: I believe that just the source folder should work but the original function had it so I opted for the better save than sorry route

**Review by wrp801 — CHANGES_REQUESTED** [review]
Why were the pyproject.toml and lock file deleted in`shipyard-bp-utils`

**Comment by johnathan-rodriguez** [comment]
I had made a prepatch version that was not needed. This was around the conversation of trying to force determine_destination_full_path to continue to work with the broken logic that was being used in the original ftp blueprint

**Comment by wrp801** [comment]
Unless I was reading it wrong, I thought it showed that you deleted those files. If you wanted to leave them alone you either need to revert that specific commit, or just copy paste the  toml from `main`

**Review by wrp801 — APPROVED** [review]
